src/upgrades/mod.rs

```diff
@@ -232,30 +444,62 @@ mod tests {
 
     use super::*;
 
-    #[test]
-    fn test_upgrade_0_8_1() {
+    async fn test_upgrade(source: PathBuf) {
         let work_dir = tmp_dir();
-        let source = PathBuf::from("test-resources/migrations/v0_8_1/");
         file::backup_dir(&source, &work_dir).unwrap();
 
-        let config = Arc::new(Config::test(&work_dir, false, false, false));
+        let config = Config::test(&work_dir, false, false, false);
         let _ = config.init_logging();
 
-        upgrade_data_to_0_9_0(config).unwrap();
+        let _upgrade = prepare_upgrade_data_migrations(UpgradeMode::PrepareOnly, Arc::new(config.clone()))
+            .await
+            .unwrap()
+            .unwrap();
+
+        // and continue - immediately, but still
+        let report = prepare_upgrade_data_migrations(UpgradeMode::PrepareThenFinalise, Arc::new(config.clone()))
+            .await
+            .unwrap()
+            .unwrap();
+
+        finalise_data_migration(report.versions(), &config).unwrap();
 
         let _ = fs::remove_dir_all(work_dir);
     }
 
-    #[test]
-    fn test_upgrade_0_6_0() {
+    #[tokio::test]
+    async fn prepare_then_upgrade_0_8_1() {
+        let source = PathBuf::from("test-resources/migrations/v0_8_1/");
+        test_upgrade(source).await;
+    }
+
+    #[tokio::test]
+    async fn prepare_then_upgrade_0_7_3_cas_only() {
+        let source = PathBuf::from("test-resources/migrations/v0_7_3_cas_only/");
+        test_upgrade(source).await;
+    }
+
+    #[tokio::test]
+    async fn prepare_then_upgrade_0_8_1_pubd_only() {
+        let source = PathBuf::from("test-resources/migrations/v0_8_1_pubd_only/");
+        test_upgrade(source).await;
+    }
+
+    #[tokio::test]
+    async fn test_upgrade_0_6_0() {
```

Comment:
Why are the other tests called `prepare_then_upgrade` but this one is not, even though this one does prepare then finalize? Ah is it because for the others you do `PrepareOnly` first and with this one you do not?

Reply:
I just wanted to hit both code paths here